### ISSUE TITLE
Faster, parallel type checking

### ISSUE THREAD

**Issue by JukkaL** [JukkaL/mypy#933]
If we have a sufficiently large program to type check we should be able to speed up type checking significantly by using multiple type checker processes that work in parallel.

One potential way to implement this:
* Have a coordinator process that maintains a module dependency graph and keeps track of a work list of modules that haven't been type checked. Initially this will contain all modules in the program.
* Have multiple worker processes that repeatedly get (sets of cyclically dependent) modules from the coordinator that have their dependencies processed, type check them and mark them as processed on the coordinator.

For this to really work we are going to need a way of efficiently communicating type check results for a module between processes (to avoid type checking shared dependencies multiple times). Having a JSON serialization format (see #932) would be sufficient.

Additionally we need a quick way of figuring out the dependency graph of all modules (or at least an approximation of it). We'll probably have to cache that between runs and update it incrementally, similar to the approach outlined in #932.

So how much would this help? Obviously this depends on the shape of the dependency graph. Under reasonable assumptions we should be able to hit an effective level of parallelism of at least 4 for larger programs, but I wouldn't be surprised if we could get even better than that. Cyclic module dependencies can add a limit to how far we can parallelize. We can probably estimate the achievable level of parallelism for a particular program by analyzing the module dependency graph.

This is probably only worth implementing after we have incremental type checking (#932), and we should preserve incrementalism -- i.e., we'd only type check modules modified since the last run and modules that depend on them.

**Comment by jhance** [comment]
Why not just manage a threadpool and whenever something is ready to be processed in a buildmanager state, it gets queued into the pool?

**Comment by jhance** [comment]
To give an idea of the scope of this, I think it would be useful to know precisely what data needs to be shared.

Is it enough to give the slave processes the semantically analyzed tree of the modules it depends on?

**Comment by JukkaL** [comment]
I explained a possible serialization format in #932. We should be able to use the same format to share data about modules in parallel builds.

A slave process could work like this, I think:

* Get a (cyclically dependent) set of files to type check from the coordinator process. For files with no cyclic dependencies the set will have just a single file. Exit if no more files are remaining in the work list at the coordinator. The get operation may block, in case the coordinator needs to wait for some other workers to finish their current jobs first.
* If none of the files are more recent than cached serialized data about the files, just do nothing and skip to the last step. Alternatively, the coordinator can do this check and only send files that need to be processed.
* Type check the files exactly like we'd do it in a single process case. In particular, process dependencies normally. We are guaranteed that the dependencies all have up-to-date serialized data available so dependencies will be quick to process. Send any type errors to the coordinator so that it can serialize output.
* Store serialized data about all the files that we just processed on the disk so that other workers can reuse our results. 
* Notify coordinator that we're finished with these files. Loop back to beginning. Maintain state so that we only have to process each dependency at most once per worker.

I'm less clear about how the coordinator would work. Some ideas:

* First incrementally update module dependency information for all modules, and cache that in separate JSON files, for example. For each file we'd maintain a list of directly imported modules. This data needs to be invalidated for all files whenever the module search path changes. Maybe we can use the standard library Python parser module to find imports in each module, as the mypy parser is pretty slow and we don't want this to become a bottleneck.
* Load and analyze the module dependency graph to at least find sets of cyclically dependent files. Again, this needs to be quick. From this point on, every unit of type checking is a set of cyclically dependent files.
* Divide units into four mutually exclusive states: 1) unprocessed units that have unprocessed dependencies, 2) unprocessed units that are ready to be type checked, 3) units being processed by a worker, 4) processed units.
* Fire up N workers.
* Send units in state 2 to workers and change them to state 3. If there are no units in state 2 but there are units in state 1, we may have to wait until a unit gets added to state 1. If there are no units in states 1 or 2, we can exit any workers that have nothing left to do.
* Once a worker finished with a unit, switch it to state 4. Potentially move some units from state 1 to 2 if they now have all their dependencies processed.  
* We are finished when every unit is in state 4.

**Comment by fzyzcjy** [comment]
Hi, is there any updates? thanks

**Comment by manmartgarc** [comment]
This would be really cool. Is there any way I can help?

**Comment by JukkaL** [comment]
@manmartgarc Any help would be very welcome! I'm happy to chat about the details over Zoom if you are interested in working on this (you can find my email in the mypy commit history).

**Comment by jgarvin** [comment]
Currently I invoke mypy on all my files like `mypy $(find . -name "*.py")`. It's starting to get a bit slow so I found this ticket. Until real support is added, is it expected to work if I run two separate mypy instances myself at the same time on different files? Then I could still parallelize at the build system level with make or similar. I realize this may lead to some redundant work but it might still come out ahead. I'm mostly concerned that the instances might conflict both trying to write to mypy's cache.

**Comment by aiven-anton** [comment]
@jgarvin In my experience, running mypy only on a part of a code-base sometimes gives different results from a full mypy run. However, I believe the intention is for this to work, and any deviation is really a bug. I assume you would need to use separate cache directories for the processes to not corrupt each other (not sure this is true though).

Also just noting, when doing a full mypy run, there should usually not be any need to filter files with `find`, you can just configure mypy with your source directory and it will traverse it.

**Comment by jgarvin** [comment]
Hmm mypy cache on a relatively small project (<20k lines) is ~40MB. So for a 128 core machine I'd be spending 4GB of disk on mypy caches which in the grand scheme of things is not huge disk space consumption but I have to imagine touching that much disk must slow down the checking. I looked at the caches to see what was responsible for the size and it seems it's a ton of JSON files which is not a very compact format. Maybe pickle would work better?

**Comment by dosisod** [comment]
@jgarvin #15731 and #15981 include some methods on reducing the size of the cached JSON files. I mentioned it in the comments somewhere, but pickling might take up more space since we only store certain fields in the JSON files, meaning pickling could potentially include data we don't need/want (have yet to look into this).

Mypy does have a SQLite cache option, though it basically just stores the JSON data and filename in a table, see https://github.com/python/mypy/issues/3456#issuecomment-630501607 .

**Comment by ilevkivskyi** [comment]
@JukkaL I started working on this. Some initial comments:
* In https://github.com/python/mypy/issues/19776 you mentioned that fast parser will help with this. IIUC however this is still viable even before we get the fast parser, right? I mean we can get some time or memory penalty for either parsing twice or copy-on-write after forking, but it would be compensated by potential 2-3x win by processing SCCs in parallel.
* It looks like we will need to change how we process files with errors, right now we don't write the cache for them _at all_ (I added this behavior LOL). Instead I think we should have something like a flag in cache meta that would force reprocessing (somewhat resembling what we do in fine-grained mode).
* To benefit the most from parallel checking, we will need to move away from current topological sort algorithm to something more "dynamic". Right now it yields ready SCCs in "batches" of mutually independent SCCs, but this is (very) inefficient when processing times for different SCCs are (very) different. Instead we want to mark an SCC ready as soon as all dependencies were processed.

I will try to split the work into multiple smaller PRs, likely starting from addressing last two items.

**Comment by JukkaL** [comment]
> IIUC however this is still viable even before we get the fast parser, right?

Yes, this should still give an improvement, but unless you can also parallelize the initial dependency DAG and SCC construction step, the maximum performance gain seems to be around 3x, and more realistically maybe 1.5x to 2x. With the fast parser, we can maybe expect typically 2x to 3x, and in ideal use cases 5x or more (when most type checked files don't depend on each other).

It seems possible to build the dependency DAG using multiple processes, but it may be not be worth it to spend the effort now if we can expect the fast parser to make this significantly easier.

> It looks like we will need to change how we process files with errors

Agreed. We could even write a full cache for files with errors if there are no syntax errors, with a list errors reported or something.

> To benefit the most from parallel checking, we will need to move away from current topological sort algorithm to something more "dynamic". 

Agreed.

**Comment by JukkaL** [comment]
> >    It looks like we will need to change how we process files with errors
>
> Agreed. We could even write a full cache for files with errors if there are no syntax errors, with a list errors reported or  something.

On reflection, we may want to report errors and the symbol table separately. We could type check module top levels and methods that assign to an attribute via `self` as a first pass, and write the cache immediately afterwards, since this is enough to determine the public interface of a module. The rest of the module could be processed afterwards as a separate type checking pass. The list of errors would only be available at the end of the second pass, but the symbol table would be available after the first pass -- and modules that depend on this module would be unblocked after the first pass. We could write the errors in a separate cache file, or perhaps include the error information in the metadata cache file.

**Comment by ilevkivskyi** [comment]
@JukkaL I will probably take this step by step. Right now not writing cache at all is a blocker. So I will start from fixing just that. Later we can make incremental improvements by publishing symbol tables earlier (which btw I think will give significant improvement).

**Comment by ilevkivskyi** [comment]
Oh well, I started from refreshing my memory on details of coarse-grained incremental mode, and already found two bugs:
* One is a crash related to indirect dependencies from a base class involving double-indirection. Although it is a crash, likely it should be easy to fix, at least I don't see any conceptual problems with it.
* Second is a stale type reported for a situation where we have a type alias and a double indirection. This one is actually problematic. An (over)simplified explanation is like this: imagine a function in module `1`, that uses an alias defined in module `3` and alias is re-exported unchanged via module `2`. The alias initially refers to a class from module `4`, but then changed to module `4'`. So we have direct dependencies `1 -> 2 -> 3 -> 4`. And an indirect dependency `1 --> 4`. Important part here is that it goes straight to `4`, not to `3`, because we call `get_proper_type()` somewhere during processing (and it is practically impossible to eliminate _all_ such calls, and prevent people from adding them). Then after update of the alias definition in `3`, module `2` is reported as interface-fresh (because we use `cross_ref` for re-exported symbols), and we get a stale type in `1` (because both dependencies `2` and `4` are reported interface-fresh).

I was thinking for a while and I have only one viable solution for the second issue: re-implement a (slightly simpler/cleaner) analog of `alias_deps` that we use for fine-grained mode (we already have them there, because apparently this problem is much more prevalent in fine-grained mode). If there are no other ideas I will make a PR today/tomorrow. (Alternatively we can simply move on with parallel checking and postpone these two issues, but I would prefer to stand on a more solid foundation before implementing any big changes).

**Comment by JukkaL** [comment]
Since a single parallel type checking run can perform the equivalent of 10+ current incremental runs, I think it makes sense to make incremental checking robust first.

**Comment by ilevkivskyi** [comment]
OK, here should be the fix https://github.com/python/mypy/pull/19798. While working on this I discovered couple other problems. I included relevant fixes in the same PR, since they are all closely related. I also looked through open incremental issues, and found half a dozen bugs worth fixing now. I will be submitting more PRs soon.

**Comment by ilevkivskyi** [comment]
OMG I have found something much worse, this passes without errors:
```
[case testIncrementalOkChangeWithSave2]
import mod1
x: int = mod1.x

[file mod1.py]
from mod2 import x

[file mod2.py]
x = 1

[file mod2.py.2]
x = "no way"
[out]
[out2]
```
And we only get away with similar story for functions because we treat `CallableType` as named type in `TypeIndirectionVisitor`. Truly one of those lolwhat moments. Unfortunately, currently the code conflates two different things together:
* The fact that we use opaque/named types (that actually requires all the sophistication of `TypeIndirectionVisitor`)
* The fact that we use symbol `cross_ref`s during symbol table serialization (that can, and probably should, be solved by attaching hash of the original symbol).

Anyway, I already have couple other incremental PRs in the pipelines, so I will get back to this after I am done with those.

**Comment by ilevkivskyi** [comment]
> that can, and probably should, be solved by attaching hash of the original symbol

On the second thought, I think the better way would be to have a "SymbolIndirectionVisitor" that would simply traverse symbol tables only, and add indirect dependencies on all foreign symbols. This would be:
* Likely more efficient.
* More conceptually consistent, we have opaque types, _and_ "opaque symbols".

**Comment by ilevkivskyi** [comment]
The more I look at this the more I am fascinated it ever worked :-) Unfortunately, it is hard to split the fixes into separate PRs, I may end up adding this to my original PR https://github.com/python/mypy/pull/19798
